### PR TITLE
sync: remove custom closing infrastructure

### PR DESCRIPTION
**Pull request by tychoish**
Contexts should be sufficient to manage the life-cycles of all these processes, so the `sync.Closer` isn't useful anymore.

### CODE REVIEW

File: internal/mempool/reactor.go [hunk 1]
```diff
@@ -53,12 +53,12 @@ type Reactor struct {
 	// goroutines.
 	peerWG sync.WaitGroup
 
+	mtx         tmsync.Mutex
+	peerClosers map[types.NodeID]context.CancelFunc
```
Comment (creachadair) [inline internal/mempool/reactor.go:57]: An option for your consideration: For locks that protect only one thing, I sometimes find this idiom useful:

```go
peerClosers struct {
   tmsync.Mutex
   m map[types.NodeID]context.CancelFunc
}

// usage:
r.peerClosers.Lock()
if c, ok := r.peerClosers.m[id]; ok { c() }
r.peerClosers.Unlock()
```

Not necessary, but makes the scope of the mutex clear. YMMV.

**Review by creachadair — APPROVED** [review]
LGTM. 

I'm not sure I understand why we sometimes use `tmsync` mutexes vs. standard `sync` package mutexes. It seems to me like we could probably drop the `tmsync` package almost entirely—not as part of this PR, but more generally.

File: internal/mempool/reactor.go (hunk 1 above)
Comment (tychoish) [reply internal/mempool/reactor.go]: yeah I've done this in the past, and like it as an idiom. 

this is just moving code around, I initially thought i could delete it, but it makes sense to leave it for now.

**Comment by tychoish** [comment]
> I'm not sure I understand why we sometimes use `tmsync` mutexes vs. standard `sync` package mutexes. It seems to me like we could probably drop the `tmsync` package almost entirely—not as part of this PR, but more generally.

Hahah, yes this is my ulterior motive but I wanted to do it in steps, and this seems more disruptive than the mutex thing (which I think is a build-time attempt to inject deadlocks into mutexes?)

File: internal/p2p/pqueue.go [hunk 2]
```diff
@@ -290,3 +274,14 @@ func (s *pqScheduler) push(pqEnv *pqEnvelope) {
 		s.sizes[uint(s.chDescs[i].Priority)] += pqEnv.size
 	}
 }
+func (s *pqScheduler) close() {
```
Comment (creachadair) [inline internal/p2p/pqueue.go:277]: Is the point here to handle double-closes gracefully? If so—a possible alternative to consider:

   signalCh <-chan struct{}
   close    func()

   sigCtx, cancel := context.WithCancel(ctx)
   // ... 
   signalCh: sigCtx.Done(),
   close:    cancel,
   
which would let `closed` work the same way and `close` would not need its own method, and multiple cancels are already handled. YMMV, though.

```suggestion

func (s *pqScheduler) close() {
```

File: internal/p2p/queue.go [hunk 3]
```diff
@@ -19,26 +17,21 @@ type queue interface {
...
-	// closed returns a channel that's closed when the scheduler is closed.
 	closed() <-chan struct{}
 }
 
 // fifoQueue is a simple unbuffered lossless queue that passes messages through
 // in the order they were received, and blocks until message is received.
 type fifoQueue struct {
 	queueCh chan Envelope
-	closer  *tmsync.Closer
+	signal  chan struct{}
```
Comment (creachadair) [inline internal/p2p/queue.go:28]: Minor nit: consistency would probably make this `signalCh`.

File: internal/p2p/transport_test.go [hunk 4]
```diff
@@ -599,7 +599,7 @@ func dialAccept(ctx context.Context, t *testing.T, a, b p2p.Transport) (p2p.Conn
 	endpoints := b.Endpoints()
 	require.NotEmpty(t, endpoints, "peer not listening on any endpoints")
 
-	ctx, cancel := context.WithTimeout(ctx, time.Second)
+	ctx, cancel := context.WithTimeout(ctx, 10*time.Second)
```
Comment (creachadair) [inline internal/p2p/transport_test.go:602]: Anything we can say here about how long this should be?  E.g.,

```go
// Leave the endpoints enough time to be reachable even if startup is slow.
```

or whatever may make sense?

**Comment by tychoish** [comment]
I keep circling around this, but I think there's not a lot of benefit to doing this (it'd be nice to delete tmsync, but it's not a huge liability) and I think there's some (very) subtle fact about the way that the behavior of closer interacts with the `MemoryTransport` implementation that's used to test the reactors and the router.

As the transport abstraction falls away with libp2p, I think it's safe (and desireable) to abandon this change set for the present/near future.

**Pull request closed without merging**